src/Plutarch/Extra/Monoid.hs

```diff
@@ -0,0 +1,63 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.Extra.Monoid (
+    PAll (..),
+    pgetAll,
+    PAny (..),
+    pgetAny,
+) where
+
+import Control.Composition ((.*))
+import Data.Function (on)
+import Plutarch (DerivePNewtype (..), PCon (pcon), PlutusType (..), S, Term, plam, type (:-->))
+import Plutarch.Bool (PBool (..), PEq, (#&&))
+import Plutarch.Builtin (PIsData)
+import Plutarch.Lift (pconstant)
+import Plutarch.Show (PShow)
+import Plutarch.Unsafe (punsafeCoerce)
+
+-- | @since 1.3.0
+newtype PAll (s :: S) = PAll (Term s PBool)
+    deriving
+        ( -- | @since 1.3.0
+          PlutusType
+        , -- | @since 1.3.0
+          PIsData
+        , -- | @since 1.3.0
+          PEq
+        , -- | @since 1.3.0
+          PShow
+        )
+        via (DerivePNewtype PAll PBool)
+
+instance forall (s :: S). Semigroup (Term s PAll) where
+    (<>) = pcon . PAll .* (#&&) `on` punsafeCoerce
```

Comment:
Do we really need/want `punsafeCoerce` here?

Reply:
Likewise for below

Reply:
It's fine to just use `punsafeCoerce` here, I suppose. Btw, using `pto` here will result in a type check error.

Reply:
I don't think this is too bad here, but I feel it'd probably be better to `pmatch` then rewrap. Newtypes are invisible at runtime.